### DISCUSSION TITLE
What is the best way to keep seed and test data in sync?

### DISCUSSION THREAD

**Discussion by Donnerstagnacht** [supabase/supabase#20082] [Questions]
I would like to keep my test data for end-to-end tests and my seed data for the database in sync to better control the test and database state.

Therefore, I thought to load the data from my seed scripts from a JSON file which I can also access in end-to-end test. But during the seed process, supabase can not find the JSON files. I described my issue in detail here:
https://stackoverflow.com/questions/77739509/is-it-possible-to-use-a-json-file-to-seed-a-supabase-postgres-database

On the supabase discord server, it was suggested that supabase can not access the file system. In that case, the approach will never work.

Is there a better way to keep the data in sync?

**Comment by Rubenburdin** [comment]
The pg_read_file approach wont work because that function runs inside the Postgres container context, not your local filesystem. Even though the migrations folder gets mounted, pg_read_file is restricted to directories Postgres explicitly has access to, like its data directory. Your notifications.json sitting next to seed.sql isnt reachable from inside the database engine.

Your inline JSON approach with json_populate_recordset is actually solid, the problem is just externalizing that JSON. A few options that work:
The cleanest is to generate seed.sql from your JSON as a build step. Write a small script that reads notifications.json and outputs the exact SQL you already have working, just with the JSON values injected inline. Run it before supabase db reset. Your e2e tests read the same JSON file directly, so its one source of truth with a tiny codegen step in between.

Another option is to seed through the Supabase client programmatically. Skip seed.sql entirely for test data and have your e2e test setup insert 
records via the API using the JSON file. Both your seed logic and your tests import the same fixtures module. Slower than raw SQL but keeps everything in one language and one place.

If you really want pure SQL, you could store your fixture JSON in a table created by migrations, then have seed.sql read from that table. A bit roundabout but it keeps the filesystem out of it entirely.
For teams dealing with keeping actual production records in sync across external systems like HubSpot or Salesforce rather than test data, thats where Stacksync comes in. But for dev fixtures the codegen approach is probably your best bet here.
